master/internal/db/postgres_experiments.go

```diff
@@ -1117,65 +1117,38 @@ WITH const AS (
-               ) AS trial_rank,
-               rank() OVER (
-                   PARTITION BY trial_id
-                   ORDER BY total_batches DESC
-               ) AS trial_order_rank
-        FROM (
-            SELECT c.id, c.trial_id, c.steps_completed as total_batches, c.state,
-                   c.report_time as end_time, c.uuid, c.resources, c.metadata,
-                   (SELECT row_to_json(s)
-                    FROM (
-                        SELECT s.end_time, s.id, s.trial_id,
-                            s.total_batches,
-                            (SELECT row_to_json(v)
-                            FROM (
-                                SELECT v.end_time, v.id, v.metrics,
-                                    v.total_batches, v.trial_id
-                                    FROM validations v
-                                    WHERE v.trial_id = t.id AND v.total_batches = s.total_batches
-                                ) v
-                               ) AS validation
-                        FROM steps s
-                        WHERE s.total_batches = c.steps_completed AND s.trial_id = c.trial_id
-                    ) s
-                   ) AS step,
-                   -- We later filter out any checkpoints with any corresponding warm start
-                   -- trials, so we can just put an empty list here. (TODO(dzhu): This is
-                   -- here for backwards compatibility with Python, but could maybe be
-                   -- removed.)
-                   '[]'::jsonb AS warm_start_trials
-            FROM checkpoints_view c, trials t, const
-            WHERE c.state = 'COMPLETED' AND c.trial_id = t.id AND t.experiment_id = $1
-        ) _, const
-    ) c, const
-    WHERE (SELECT COUNT(*) FROM trials t WHERE t.warm_start_checkpoint_id = c.id) = 0
-          AND c.trial_order_rank > $4
-          AND ((c.experiment_rank > $2
-                AND c.trial_rank > $3)
-               OR (c.step->'validation'->'metrics'->'validation_metrics'->>const.metric_name
-                   IS NULL))
+	SELECT c.uuid,
+		-- The order includes the id to prevent different rows from having the same
+		-- rank, which could cause more than the desired number of checkpoints to be
+		-- left out of the result set. Also, any rows with null validation values
+		-- will sort to the end, thereby not affecting the ranks of rows with
+		-- non-null validations, and will be filtered out later.
+		rank() OVER (
+			ORDER BY const.sign * (v.metrics->'validation_metrics'->>const.metric_name)::float8
+			ASC NULLS LAST, v.id ASC
+		) AS experiment_rank,
+		rank() OVER (
+			PARTITION BY trial_id
+			ORDER BY const.sign * (v.metrics->'validation_metrics'->>const.metric_name)::float8 
+			ASC NULLS LAST, v.id ASC
+		) AS trial_rank,
+		rank() OVER (
+			PARTITION BY trial_id
+			ORDER BY (c.metadata->>'steps_completed')::int DESC
+		) AS trial_order_rank
+	FROM checkpoints_v2 c
+	NATURAL JOIN const
```

Comment:
i know i helped write this but it is probably clearer as
```suggestion
	JOIN const ON true
```

Reply:
i played around with doing it with `CROSS JOIN` and `LEFT JOIN LATERAL` for fun. all possible, all emit similar plans and this is the most clear to me.

Reply:
can we go ahead and pull this in?

Reply:
oh we merged.